internal/cmd/gocdk/launch.go

```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
 	return serveURL, nil
 }
 
+// cloudRunLauncher pushes Docker containers to Google Container Registry and
+// creates/updates a Cloud Run service.
+type cloudRunLauncher struct {
+	logger    *log.Logger
+	client    *cloudrun.APIService
+	dockerEnv []string
+	dockerDir string
+}
+
+// Launch implements Launcher.Launch.
+func (crl *cloudRunLauncher) Launch(ctx context.Context, input *LaunchInput) (*url.URL, error) {
+	projectID := specifierStringValue(input.Specifier, "project_id")
+	location := specifierStringValue(input.Specifier, "location")
+	serviceName := specifierStringValue(input.Specifier, "service_name")
+	if projectID == "" || location == "" || serviceName == "" {
+		return nil, xerrors.New("cloud run launch: specifier does include project_id, location, and/or service_name")
+	}
+
+	// Push to GCR if needed.
+	// TODO(light): Check for presence in Docker daemon and country-specific
+	// prefixes.
+	if strings.HasPrefix(input.DockerImage, "gcr.io/") {
+		if err := crl.push(ctx, input.DockerImage); err != nil {
+			return nil, xerrors.Errorf("cloud run launch: %w", err)
+		}
+	}
+
+	// Launch on Cloud Run.
+	var env []*cloudrun.EnvVar
+	for i, inputVar := range input.Env {
+		eqIdx := strings.IndexByte(inputVar, '=')
+		if eqIdx == -1 {
+			return nil, xerrors.Errorf("cloud run launch: bad format for specifier env[%d] (got %q)", i, inputVar)
```

Comment:
Same here about being more specific about where this is coming from. And tell them what the format should be, not just that it's bad.

Reply:
Done.